Spigot-Server-Patches/0307-Vanished-players-don-t-have-rights.patch

```diff
@@ -0,0 +1,131 @@
+     }
+ 
++    public boolean isEmpty() { return b(); } // Paper - OBFHELPER
++
+     public VoxelShape a(double d0, double d1, double d2) {
+         return (VoxelShape) (this.b() ? VoxelShapes.a() : new VoxelShapeArray(this.a, new DoubleListOffset(this.a(EnumDirection.EnumAxis.X), d0), new DoubleListOffset(this.a(EnumDirection.EnumAxis.Y), d1), new DoubleListOffset(this.a(EnumDirection.EnumAxis.Z), d2)));
+     }
+diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
+index 127dcedc..302a3655 100644
+--- a/src/main/java/net/minecraft/server/World.java
++++ b/src/main/java/net/minecraft/server/World.java
+@@ -1544,6 +1544,37 @@ public abstract class World implements GeneratorAccess, IIBlockAccess, AutoClose
+         }
+     }
+ 
++    // Paper start - Based on method below
++    /**
++     * @param entity causing the action ex. block placer
++     * @param voxelshape area to search within
++     * @return if there are no visible players colliding
++     */
++    public boolean checkNoVisiblePlayerCollisions(@Nullable Entity entity, VoxelShape voxelshape) {
++        if (voxelshape.isEmpty()) {
++            return true;
++        } else {
++            List list = this.getEntities((Entity) null, voxelshape.getBounds());
++
++            for (int i = 0; i < list.size(); ++i) {
++                Entity entity1 = (Entity) list.get(i);
++
++                if (entity instanceof EntityPlayer && entity1 instanceof EntityPlayer) {
++                    if (!((EntityPlayer) entity).getBukkitEntity().canSee(((EntityPlayer) entity1).getBukkitEntity())) {
++                        continue;
++                    }
++                }
++
++                if (!entity1.dead && entity1.blocksEntitySpawning()) {
++                    return false;
++                }
++            }
++
++            return true;
++        }
++    }
++    // Paper end
++
+     public boolean a(@Nullable Entity entity, VoxelShape voxelshape) {
+         if (voxelshape.b()) {
+             return true;
+diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+index cf398cd2..99da4c22 100644
+--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
++++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+@@ -804,6 +804,13 @@ public class CraftEventFactory {
+         Projectile projectile = (Projectile) entity.getBukkitEntity();
+         org.bukkit.entity.Entity collided = position.entity.getBukkitEntity();
+         com.destroystokyo.paper.event.entity.ProjectileCollideEvent event = new com.destroystokyo.paper.event.entity.ProjectileCollideEvent(projectile, collided);
++
++        if (projectile.getShooter() instanceof Player && collided instanceof Player) {
++            if (!((Player) projectile.getShooter()).canSee((Player) collided)) {
++                event.setCancelled(true);
```

Comment:
Go ahead and return here. We shouldn't fire the event. There should not be opportunity for a plugin to break this.

Reply:
I know original patch didn't do that, time to improve ;)